src/chainstate/coordinator/mod.rs

```diff
@@ -658,6 +737,240 @@ impl<
+        stacks_block_id: &StacksBlockId,
+    ) -> Result<bool, Error> {
+        let reward_cycle_length = self.burnchain.pox_constants.reward_cycle_length;
+        let veto_percent_threshold = self
+            .burnchain
+            .exit_contract_constants
+            .veto_confirmation_percent;
+        let is_mainnet = self.burnchain.is_mainnet();
+        self.chain_state_db
+            .with_read_only_clarity_tx(&self.sortition_db.index_conn(), &stacks_block_id, |conn| {
+                conn.with_clarity_db_readonly(|db| {
+                    // TODO (#3034): get correct contract ID once contract is published
+                    let exit_at_rc_contract = boot_code_id("exit-at-rc", is_mainnet);
+
+                    // from map rc-proposal-vetoes, use key pair (proposed_rc, curr_rc) to get the # of vetos
+                    let entry = db
+                        .fetch_entry_unknown_descriptor(
+                            &exit_at_rc_contract,
+                            "rc-proposal-vetoes",
+                            &Value::from(
+                                TupleData::from_data(vec![
+                                    ("proposed-rc".into(), Value::UInt(proposed_exit_rc as u128)),
+                                    ("curr-rc".into(), Value::UInt(rc_cycle_of_veto as u128)),
+                                ])
+                                .expect("BUG: failed to construct simple tuple"),
+                            ),
+                        )
+                        .expect("BUG: Failed querying rc-proposal-vetoes")
+                        .expect_optional();
+
+                    let num_vetos = match entry {
+                        Some(val) => {
+                            let tuple_data = val.expect_tuple();
+                            tuple_data
+                                .get("vetoes")
+                                .expect("BUG: malformed cost proposal tuple")
+                                .clone()
+                                .expect_u128()
+                        }
+                        None => 0,
+                    };
+
+                    println!(
+                        "RCPR: in veto check, proposed_exit: {}, num vetos: {:?}",
+                        proposed_exit_rc, num_vetos
+                    );
+                    // Check if the percent veto crosses the minimum threshold
+                    let percent_veto = num_vetos * 100 / (reward_cycle_length as u128);
+
+                    Ok(percent_veto >= (veto_percent_threshold as u128))
+                })
+            })
+            .ok_or(Error::DBError(NotFoundError))?
+    }
+
+    /// Returns map of exit proposals to the number of votes for each, as well as the sum total of all
+    /// votes.
+    pub fn read_vote_state(
+        &mut self,
+        rc_cycle_of_vote: u64,
+        curr_exit_at_rc_opt: Option<u64>,
+        stacks_block_id: &StacksBlockId,
+        invalid_reward_cycles: Vec<u64>,
```

Comment:
If this list is bound in length, then I think it should be a slice of `Option<u64>` (where `Some(..)` indicates that the entry should be considered).

Reply:
Just to confirm I know what you are suggesting, let's say the range of possible exit proposals is 15 to 30, and the invalid reward cycles are (18, 21). 
Would I have a slice of length 15, that looks like [Some(18), Some(21), None, None, None, None, ...]?

Reply:
@jcnelson

Reply:
Yes, but on a subsequent pass, I'm not sure that the precondition ("this list is bound in length") is met in the first place.  If that's the case, then what you have here is fine.